caikit/runtime/work_management/abortable_context.py

```diff
@@ -0,0 +1,178 @@
+
+    @abc.abstractmethod
+    def set_context(self, context: AbortableContextBase):
+        """Set the abortable context that must be notified to abort work"""
+
+    @abc.abstractmethod
+    def unset_context(self):
+        """Unset any abortable context already held. Do not notify it that work should abort"""
+
+
+class ThreadInterrupter:
+    """This class implements a listener which will observe all ongoing work in `AbortableContexts`
+     and raise exceptions in the working threads if they need to be aborted.
+
+    This offers a performance advantage over the old `AbortableActions`, since only one extra
+    listener thread is created that lives for the whole lifetime of the program. The caveat
+    is that all work to be aborted must be running in a thread that is safe to kill: this may
+    not be safe to use with asyncio tasks running in an event loop.
+    """
+
+    def __init__(self):
+        # Using a SimpleQueue because we don't need the Queue's task api
+        self._queue = SimpleQueue()
+        self._thread = None
+
+        self._context_thread_map: Dict[uuid.UUID, int] = {}
+
+        self._shutdown = True
+
+    def start(self):
+        if not self._shutdown:
+            log.debug("ThreadInterrupter already started")
+            return
+        log.debug("Starting ThreadInterrupter")
+        self._shutdown = False
+        self._thread = threading.Thread(target=self._watch_loop)
+        self._thread.start()
+
+    def stop(self):
+        if self._shutdown:
+            log.debug("ThreadInterrupter already shut down")
+            return
+
+        log.info("Stopping ThreadInterrupter")
+        self._shutdown = True
+        self._queue.put(0)
+        self._thread.join(timeout=1)
+
+    def register(self, context_id: uuid, thread: int) -> None:
+        self._context_thread_map[context_id] = thread
+
+    def unregister(self, context_id: uuid) -> None:
+        self._context_thread_map.pop(context_id, None)
+
+    def kill(self, context_id: uuid) -> None:
+        # Put this context onto the queue for abortion and immediately return
+        self._queue.put(context_id, block=False)
+
+    def _watch_loop(self):
+        while True:
+            try:
+                log.debug4("Waiting on any work to abort")
+                context_id = self._queue.get()
+
+                if self._shutdown:
+                    log.debug4("Ending abort watch loop")
+                    return
+
+                self._kill_thread(context_id)
+
+                # Ensure this context/thread pair is unregistered
+                self.unregister(context_id)
+
+            except Exception:
+                log.warning("Caught exception while running abort loop", exc_info=True)
+
+    def _kill_thread(self, context_id: uuid.UUID) -> bool:
+        thread_id = self._context_thread_map.get(context_id, None)
+
+        if thread_id:
+            async_exception_result = ctypes.pythonapi.PyThreadState_SetAsyncExc(
```

Comment:
I remember there being a really nice big comment about this magic in `AbortableAction`. Can we port that over here?

Reply:
ah true